arch/arm/src/common/arm_semi_syslog.c

```diff
@@ -54,15 +52,25 @@ int up_putc(int ch)
 }
 
 /****************************************************************************
- * Name: up_puts
+ * Name: up_nputs
  *
  * Description:
  *   Output a string on the console
  *
  ****************************************************************************/
 
-void up_puts(const char *str)
+void up_nputs(const char *str, size_t len)
 {
-  smh_call(SEMI_SYSLOG_WRITE0, (char *)str);
+  if (len == ~((size_t)0))
+    {
+      smh_call(SEMI_SYSLOG_WRITE0, (char *)str);
```

Comment:
Should we equip `SEMI_SYSLOG_WRITE0` with additional parameter `len`?

Reply:
SEMI_SYSLOG_WRITE0 is defined by ARM, we can't change it as we want to:
https://github.com/ARM-software/abi-aa/blob/main/semihosting/semihosting.rst#sys-write0-0x04
since the PC tool will stop working.

Reply:
Understood

Reply:
Yes, but in your changes you assume that string might not be null-terminated, so both `str` and `len > 0` are performed. According to the description that you pointed the WRITE0 is: `Writes a null-terminated string to the debug channel.`, so the ARM debug interface might become broken. 
I mean that if we know that string is null-terminated then `len > 0` check may be omitted in the default implementation of `up_nputs`.

Reply:
I see... Only `~0` is for WRITE0 call and means that string is null-terminated.
So indeed, my previous comment about `*str` check in else was wrong. In the else case we know the exact length, so `*str` can be omited

Reply:
Yes, done.